OffchainLabs/arbitrum-token-bridge
# feat: add support for adding custom chains

Requires https://github.com/OffchainLabs/arbitrum-sdk/pull/312
Closes https://github.com/OffchainLabs/arbitrum-token-bridge/issues/1046

### Background
To fully support Orbit chains, we need to allow users to integrate custom chains with our UI. This feature will do that by utilizing new SDK method: `addCustomChain`.

It works similar to `addCustomNetwork`, but instead of taking `customL1Network` and `customL2Network` it takes `customParentChain` and `customChain`. Then instead of adding networks to `l1Networks` and `l2Networks`, it adds them to `parentChains` and `chains`.
 
The difference between `L1Network` vs `ParentChain` & `L2Network` vs `Chain`:
1. `L1Network` should be clear, any Ethereum network. `ParentChain` is any Ethereum network OR any Arbitrum network that has at least one `partnerChainIDs` element in the array. If it does, it means it's a parent chain to some network - this 'child' network can be either an L2 or an Orbit chain (L1 <> L2 or L2 <> Orbit).
2. `L2Network` also clear, any Arbitrum network on Layer 2. `Chain` is any L2 network OR an Orbit chain. `Chain` can be thought of as a child chain - they always have a corresponding `ParentChain`.

### What was done
We basically allow users to add a new custom `Chain`. They cannot add a custom `ParentChain` to prevent phishing. Instead, they specify it in the `Chain` object with `partnerChainID`. It has to be either an Ethereum testnet or an Arbitrum testnet. We don't pass the actual `partnerChainID` network as `customParentChain` because it already exists in the network list.

### How to test
[SDK changes](https://github.com/OffchainLabs/arbitrum-sdk/pull/312) are being reviewed. Therefore, you have to use `yarn link` to test the UI. Checkout [this branch](https://github.com/OffchainLabs/arbitrum-sdk/tree/support-l2-parent-chains) and then run: `yarn build` followed by `yarn link`.

Then, inside the token bridge repo run: `cd packages/arb-token-bridge-ui/` followed by `yarn link "@arbitrum/sdk`. Now you should be able to test the UI.

You can find this feature in *Preferences*. To test, try the following JSON configs. `chainID` and `name` don't matter, all we care about is `partnerChainID`.
1. `{ "chainID": "12345", "name": "Network 2", "partnerChainID": "5" }` - 5 is Goerli, which means it's a valid Ethereum testnet (remember only an Ethereum testnet or an Arbitrum testnet is allowed).
2. `{ "chainID": "12345", "name": "Network 2", "partnerChainID": "1" }` - 1 is Mainnet, which is not allowed. It should throw an error.

You can try out different things to break it. Let me know if you have any questions.
